File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

import java.util.Objects;

public class Animal{
    private MapDirection orientation = MapDirection.NORTH;
    private Vector2d position;

    public Animal(Vector2d position) {
        this.position = position;
    }

    public Animal() {
        this(new Vector2d(2, 2));
    }

    public String toString() {
        return orientation.toString();
    }

    public boolean isAt(Vector2d position) {
        return this.position.equals(position);
    }

    public void move(MoveDirection direction, MoveValidator<Vector2d> validator) {
        switch (direction) {
            case RIGHT -> orientation = orientation.next();
            case LEFT -> orientation = orientation.previous();
            case FORWARD -> {
                Vector2d newPosition = position.add(orientation.toUnitVector());
                if (validator.canMoveTo(newPosition)) {
                    position = newPosition;
                }
            }
            case BACK -> {
                Vector2d newPosition = position.subtract(orientation.toUnitVector());
                if (validator.canMoveTo(newPosition)) {
                    position = newPosition;
                }
            }
        }
    }

    public MapDirection getOrientation() {
        return orientation;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Animal animal = (Animal) o;
        return orientation == animal.orientation && Objects.equals(position, animal.position);
    }

    @Override
    public int hashCode() {
        return Objects.hash(orientation, position);
    }

    public Vector2d getPosition() {
        return position;
    }
}

File: Kotlin/src/main/kotlin/Animal.kt
data class Animal(private val initialPosition: Vector2d = Vector2d(2, 2)) {
    private var orientation: MapDirection = MapDirection.NORTH
    private var position: Vector2d = initialPosition

    override fun toString(): String {
        return orientation.toString()
    }

    fun isAt(position: Vector2d): Boolean {
        return this.position == position
    }

    fun move(direction: MoveDirection, validator: MoveValidator<Vector2d>) {
        when (direction) {
            MoveDirection.RIGHT -> orientation = orientation.next()
            MoveDirection.LEFT -> orientation = orientation.previous()
            MoveDirection.FORWARD -> {
                val newPosition = position + orientation.toUnitVector()
                if (validator.canMoveTo(newPosition)) {
                    position = newPosition
                }
            }
            MoveDirection.BACK -> {
                val newPosition = position - orientation.toUnitVector()
                if (validator.canMoveTo(newPosition)) {
                    position = newPosition
                }
            }
        }
    }

    fun getOrientation(): MapDirection {
        return orientation
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Animal) return false

        return orientation == other.orientation && position == other.position
    }

    override fun hashCode(): Int {
        return listOf(orientation, position).hashCode()
    }

    fun getPosition(): Vector2d {
        return position
    }

    fun setPosition(newPosition: Vector2d) {
        position = newPosition
    }
}